src/systems/app-mode.js

```diff
@@ -0,0 +1,227 @@
+    // TODO: come up with better huristics for this that maybe account for the user turning away from the hud "too far", also animate the position so that it doesnt just snap.
+    if (yawDif >= yawCutoff || pitch > lookCutoff + animRange / 3) {
+      const lookDir = new THREE.Vector3(0, 0, -1);
+      lookDir.applyQuaternion(head.quaternion);
+      lookDir.add(head.position);
+      hud.position.x = lookDir.x;
+      hud.position.z = lookDir.z;
+      hud.setRotationFromEuler(new THREE.Euler(0, head.rotation.y, 0));
+    }
+
+    //animate the hud into place over animRange degrees as the user aproaches the lookCutoff angle
+    const t = 1 - THREE.Math.clamp(pitch - lookCutoff, 0, animRange) / animRange;
+
+    // Lock the hud in place relative to a known head position so it doesn't bob up and down
+    // with the user's head
+    if (!this.isYLocked && t === 1) {
+      this.lockedHeadPositionY = head.position.y;
+    }
+    const EPSILON = 0.001;
+    this.isYLocked = t > 1 - EPSILON;
+
+    hud.position.y = (this.isYLocked ? this.lockedHeadPositionY : head.position.y) - offset - offset * (1 - t);
+    hud.rotation.x = (1 - t) * THREE.Math.DEG2RAD * 90;
+
+    // update the app mode when the HUD locks on or off
+    // TODO: this assumes full control over current app mode reguardless of what else might be manipulating it, this is obviously wrong
+    const AppModeSystem = sceneEl.systems["app-mode"];
+    if (pitch < lookCutoff && AppModeSystem.mode !== AppModes.HUD) {
+      AppModeSystem.setMode(AppModes.HUD);
+      sceneEl.renderer.sortObjects = true;
+    } else if (pitch > lookCutoff && AppModeSystem.mode === AppModes.HUD) {
+      AppModeSystem.setMode(AppModes.DEFAULT);
+      sceneEl.renderer.sortObjects = false;
+    }
+  }
+});
+
+/**
+ * Toggle visibility of an entity based on if the user is in vr mode or not
+ */
+AFRAME.registerComponent("vr-mode-toggle-visibility", {
+  schema: {
+    invert: { type: "boolean", default: false }
+  },
+
+  init() {
+    this.updateComponentState = this.updateComponentState.bind(this);
+    this.updateComponentState();
+  },
+
+  play() {
+    this.el.sceneEl.addEventListener("enter-vr", this.updateComponentState);
+    this.el.sceneEl.addEventListener("exit-vr", this.updateComponentState);
+  },
+
+  pause() {
+    this.el.sceneEl.removeEventListener("enter-vr", this.updateComponentState);
+    this.el.sceneEl.removeEventListener("exit-vr", this.updateComponentState);
+  },
+
+  updateComponentState(i) {
+    const inVRMode = this.el.sceneEl.is("vr-mode");
+    this.el.setAttribute("visible", inVRMode !== this.data.invert);
+  }
+});
+
+/**
+ * Toggle the isPlaying state of a component based on app mode
+ */
+AFRAME.registerComponent("vr-mode-toggle-playing", {
+  multiple: true,
+  schema: {
+    invert: { type: "boolean", default: false }
+  },
+
+  init() {
+    this.updateComponentState = this.updateComponentState.bind(this);
+    this.updateComponentState();

```

Comment:
This should be in `play()`, since A-frame will play all the components when the scene has finished init'ing.

Reply:
✔️